README.md
# Perpetually Against Humanity, IRC Edition (pah-irc)

## What?

It's another IRC bot for playing [Cards Against
Humanity™](http://www.cardsagainsthumanity.com/). Cards Against Humanity is a
trademark of Cards Against Humanity LLC.

[Wikipedia](http://en.wikipedia.org/wiki/Cards_Against_Humanity) describes it
as:

> a party game using cards. It is available as a free download that players can
> print to create their own cards, and also available to purchase in published
> hardcopy.

Here's [some logs from a real game of Perpetually Against
Humanity](http://grifferz.github.io/pah-irc/doc/logs.html).

## Why?

There's quite a few IRC bots for this already, but it struck me that none of
them particularly played to IRC's strengths, those being:

*   Asynchronous, long-term associations between a slowly evolving group

    IRC channels tend to exist for a long time with a group of people that
    alters quite slowly. Your typical real-life game of CAH involves a bunch of
    people at a specific event for a limited amount of time. An IRC channel may
    exist for years.

    So why can't you play CAH for years, but slowly?

    And why not have the players able to come and go? That's a disaster for an
    in-person limited-time game, but not a big deal for something that's
    essentially going to go on forever. The bot should know whose turn it is to
    do what, and remind them when it next sees them.

*   Opportunity for more natural interaction with a program

    IRC bots don't have to be incredibly terse using commands with tons of
    syntax.  It should be possible to conduct a simple game using simple
    English language expressions on the part of the players.

*   Multiple things happening at once

    No reason why the bot can't be running games in multiple channels at once.
    Running extra copies is going to be a bit tedious.

## Usage

Quite a few interactions with the bot will be fairly natural language and
depend on context, for example:

> &lt;AgainstHumanity&gt; We need 3 more players to get this game started. Who
                          else wants to play? Say "AgainstHumanity: me" if
                          you'd like to join in.

> &lt;grifferz&gt; AgainstHumanity, me

> &lt;AgainstHumanity&gt; Great, you're in! 2 more players needed…

Which hopefully doesn't need to be documented.

### Public channel commands

Public channel commands should be addressed to the bot by just sending a
message to the channel with the bot's nickname as the first word. In the
documentation we'll assume that the bot's nickname is **AgainstHumanity**, so
for example:

> AgainstHumanity status

> AgainstHumanity: status

> AgainstHumanity, status

are all fine.

*   `status`

    Report the status of the game currently playing in the channel, if any.

    If a game is currently playing then this will say:

    *   Who we're waiting on (either to play their turn, or to pick the winner
        if they're the Card Tsar);

    *   Who the current Card Tsar is.

    *   The text of the current Black Card;

*   `scores`

    `stats`

    Displays:

    *   Who's active in the game and what their scores are.

    *   The top three all-time points-scorers for this game, including ties.

*   `start`

    Starts a new game of Perpetually Against Humanity. Players will need to be
    gathered before it will actually proceed.

*   `dealmein`

    `deal me in`

    Join in to the currently-active game. You'll immediately receive a hand and
    a reminder of the current Black Card by private message. You can play your
    hand immediately.

    If the game already has 20 players then you will need to wait until
    one of them resigns.

    If the game now has enough players to start then it will start.

*   `resign`

    `dealmeout`

    `deal me out`

    `retire`

    `quit`

    Resign from the currently-active game. Your scores will still be kept and
    you can join in to the game again later on. If this takes the number of
    active players below 4 then the game will be paused until someone joins
    again.

    Your hand of White Cards including any White Cards you already played are
    discarded.

    If you are the current round's Card Tsar then play will move on without any
    Awesome Points being awarded. If the hand was complete (all players made
    their play and plays were announced in the channel) then these White Cards
    will be discarded and new White Cards drawn. Otherwise played White Cards
    will be returned to players' hands.

*   `winner <number>`

    `<number>`

    Used by the Card Tsar to indicate that the play numbered <number> is the
    one they think is best.

    This command can be simply abbreviated as a single digit, as it will be
    obvious from the context what is meant.

*   `plays`

    Repeat the list of plays for the completed hand, if for some reason you
    missed it.

### Private commands

These commands can be given in private message to the bot. Again there are some
interactions that are more natural, in response to the state of the game.

Many commands optionally take a channel parameter. This is only needed if the
player is in more than one active game. That's fairly unlikely so most of the
time that can be left off and the bot will work it out.

*   `[#channel] status`

    Identical to the public `status` command, but in private in case you don't
    want to bother the channel with the output.

*   `[#channel] scores`

    `[#channel] stats`

    Identical to the public `scores` command, but in private in case you don't
    want to bother the channel with the output.

*   `[#channel] black`

    Repeat the text of the current Black Card.

*   `[#channel] hand`

    `[#channel] list`

    Provides a numbered list of the White Cards in the player's hand.

*   `[#channel] play <number>`

    `[#channel] play <number> <number>`

    `[#channel] play <number>,<number>`

    `[#channel] play <number> and <number>`

    `[#channel] play <number> & <number>`

    `<number>`

    `<number> <number>`

    Play the numbered White Card for this round. Some rounds require two White
    Cards, so they will be played in the order specified.

    The bot will repeat back to you the current Black Card and your proposed
    answer.

    You may alter your play as many times as you like until every active player
    has made their play, at which point the plays will be revealed to the
    channel and may not be changed.

    Where a channel will not be required, just saying one or two numbers
    (without the "play" prefix) is an acceptable shortcut.

*   `plays`

    Repeat the list of plays, if for some reason you missed it.

*   `[#channel] deck`

    Returns information about the deck in use:

    * List of card packs of which it is comprised.

    * Total Black and White Card count.

    * How many Black/White cards are left before reshuffle.

*   `config [key] [value]`

    `settings [key] [value]`

    `setting [key] [value]`

    `set [key] [value]`

    View or change a personal configuration setting. With no arguments, this
    displays the current value of all configuration settings. Specify a key to
    view just that key. Specify a key and a value to change the key to that
    value.

    Available configuration:

    * `chatpoke [Yes|1|On|No|0|Off]`

      By default if the bot sees you chatting and knows you have yet to do your
      duty it will send you a message to poke you in to action. If you don't
      like that then you can turn it off here.

    * `pronoun <pronoun>`

      By default the bot uses the "their" pronoun:

      * `<PAH>` dutchie: A game is active! We're just waiting on **dave2** to make their play.
      * […]
      * `<PAH>` 3 people just played! We're just waiting on **dg** to make their play.

      If you would prefer for the bot to use a different pronoun in relation to
      you, set it here. This will apply in all games that you are in.

## Installation

You can find the dependencies in the **cpanfile**, but you may find it simpler
to install them all into a local directory with **cpanminus**:

```
$ cpanm --local-lib=./pah-libs --installdeps .
```

Or if you'd rather use CPAN modules for everything that's not core Perl:

```
$ cpanm --local-lib-contained=./pah-libs --installdeps .
```

### IRC nickname

Now is a good time to register an IRC nickname for the bot to use. It can help
with flooding to also make sure the bot is voiced in channels it's likely to be
used in, but it will still work if not.

### Configuration file

The example configuration file at `etc/pah-irc.conf.example` should contain
enough comments and the default values for you to be able to work this out.
Just copy it to `etc/pah-irc.conf` and then edit it to suit.

### Database

The bot currently uses an SQLite database file. I've tried to use unexciting
SQL though so it should be simple to make it work with other DB engines. Please
submit an issue if you need that.

Before your first run of the bot you'll need to create the database like so:

```bash
$ bin/upgrade_db
DBIx::Class::Schema::Versioned::_on_connect(): Your DB is currently unversioned. Please call upgrade on your schema to sync the DB. at bin/upgrade_db line 41
```

An empty [SQLite](http://www.sqlite.org/) 3 database file will then exist at
`var/pah-irc.sqlite`, which you can examine with:

```bash
$ sqlite3 var/pah-irc.sqlite '.tables'
bcards                      users_games               
channels                    users_games_discards      

CHANGELOG.md
# Change log

## v0.7
### 2015-04-28
* Added pack of White Cards from main Cards Against Humanity US edition (Douglas Gardner): [#134](https://github.com/grifferz/pah-irc/issues/134).
* Added new `config` command and moved `pronoun` command under it (more configuration to come): [#113](https://github.com/grifferz/pah-irc/issues/113).
* Added new `deck` command which gives information about the packs which comprise the deck: [#129](https://github.com/grifferz/pah-irc/issues/129).
* Fixed bug where White Cards were not being shuffled properly: [#141](https://github.com/grifferz/pah-irc/issues/141).
* Fixed bug where multiple resignations during a completed round (i.e. when all plays are in) could extend the round timer by a full `turnclock` for each one: [#130](https://github.com/grifferz/pah-irc/issues/141).
* Fixed bug where card packs being returned in hash order rather than configured order was causing spurious deck rebuilds: [#140](https://github.com/grifferz/pah-irc/issues/140).
* Fixed occasional broken punctuation when Black Card had no blanks: [#8](https://github.com/grifferz/pah-irc/issues/8).
* Fixed bug where players waiting to join would end up with an empty hand if the game was progressed due to resignation of the Card Tsar: [#143](https://github.com/grifferz/pah-irc/issues/143).
* Fixed bug where new games had a discrepancy between configured pack names and database pack names: [#142](https://github.com/grifferz/pah-irc/issues/142).
* Typo fixes:
    * (Jon Fautley): [#144](https://github.com/grifferz/pah-irc/pull/144).
    * (Josh Holland): [#133](https://github.com/grifferz/pah-irc/pull/133).
    * [#135](https://github.com/grifferz/pah-irc/issues/135).

## v0.6
### 2015-04-04
* Cards Against Humanity Holiday 2013 Expansion added (Douglas Gardner): [#127](https://github.com/grifferz/pah-irc/pull/127).
* Cards Against Humanity 3rd Expansion, PAX East and PAX Prime added (Douglas Gardner): [#123](https://github.com/grifferz/pah-irc/pull/123).
* Added Lugradio-themed card pack from Bruno Bord, though it will need work before it's usable: [#120](https://github.com/grifferz/pah-irc/issues/120).
* Bot now checks it's in the correct channels whenever it has identified to its nickname and again periodically. This solves the problem where it tries to join channels before it's identified to its nick, fails for those that require a registered nickname, and never tries again: [#122](https://github.com/grifferz/pah-irc/issues/122).
* Bot is now able to kick someone off that is using its nick, and take its nick back: [#94](https://github.com/grifferz/pah-irc/issues/94).
* Added `quit` command as an alias for `resign`: [#119](https://github.com/grifferz/pah-irc/issues/119).
* Batched notification of plays made now happens after a maximum of 30 minutes. Previously this was `turnclock` divided by 60 (24 minutes per day of `turnclock`), or 72 minutes for a game with a `turnclock` of 3 days, which was felt to be too long: [#117](https://github.com/grifferz/pah-irc/issues/117).
* Bot now asks ChanServ for voice whenever it joins a channel, as having the bot voiced is required to avoid flood-limiting: [#116](https://github.com/grifferz/pah-irc/issues/116).
* Players who try to join the game while it's waiting for the Tsar to pick the winner are now automatically added at the next round, instead of being told to try again later: [#39](https://github.com/grifferz/pah-irc/issues/39).
* Fixed bug where someone resigning after the round has already been completed could potentially extend the round time out by another full `turnclock`: [#126](https://github.com/grifferz/pah-irc/issues/126).
* Fixed cosmetic bug where nothing would appear to happen when a non-player tried to pick the winner: [#121](https://github.com/grifferz/pah-irc/issues/121).
* Fixed minor bug in the `status` command where it said you were already playing even if you weren't: [#112](https://github.com/grifferz/pah-irc/issues/112).
* Fixed some more places where a lowercased version of a player's nickname would be used: [#101](https://github.com/grifferz/pah-irc/issues/101).

## v0.5
### 2015-03-22
* Implemented support for multiple card packs (currently a bot-wide setting): [#75](https://github.com/grifferz/pah-irc/issues/75).
* Added `plays` public and private commands to repeat the list of played cards for the completed hand: [#10](https://github.com/grifferz/pah-irc/issues/10).
* Fixed bug where changing your play would cause the wrong number of plays to be reported: [#89](https://github.com/grifferz/pah-irc/issues/89).

lib/PAH/Command/Pub.pm
package PAH::Command::Pub;

=pod
The commands that can be received from IRC in a public channel.

Copyright ©2015 Andy Smith <andy-pah-irc@strugglers.net>

Artistic license same as Perl.
=cut

use utf8;

use PAH::Log;

sub scores {
    my ($self, $args) = @_;

    my $chan    = $args->{chan};
    my $who     = $args->{nick};
    my $schema  = $self->_schema;
    my $irc     = $self->_irc;

    my $channel = $self->db_get_channel($chan);

    # It shouldn't be possible to not have a Channel row, because we wouldn't
    # be inside the channel if we didn't know about it.
    if (not defined $channel) {
        $irc->msg($chan,
            "$who: Sorry, I don't seem to have $chan in my database, which is"
            . " a weird error that needs to be reported!");
        return;
    }

    my $game = $channel->rel_game;

    # How long ago did we last do this?
    my $now = time();

    if (defined $game and defined $self->_last->{$game->id}
            and defined $self->_last->{$game->id}->{scores}) {
        my $last_scores = $self->_last->{$game->id}->{scores};

        if (($now - $last_scores) <= 60) {
        # Last time we did scores in this channel was 60 seconds ago or less.
            debug("%s tried to display scores for %s but it was already done %u"
                . " secs ago; ignoring", $who, $chan, ($now - $last_scores));
            $irc->msg($chan,
                sprintf("$who: Sorry, I'm ignoring your scores command"
                    . " because I did one just %u secs ago.",
                    ($now - $last_scores)));
            return;
        }

        # Record timestamp of when we did this.
        $self->_last->{$game->id}->{scores} = $now;
    }

    if (not defined $game) {
        # There's never been a game in this channel.
        $self->no_such_game($chan, $chan, $who);
        return;
    }

    # Game is either running or gathering players.
    $self->report_game_scores($game, $chan);
}

sub status {
    my ($self, $args) = @_;

    my $chan   = $args->{chan};
    my $who    = $args->{nick};
    my $schema = $self->_schema;
    my $irc    = $self->_irc;

    my $channel = $self->db_get_channel($chan);

    # It shouldn't be possible to not have a Channel row, because we wouldn't
    # be inside the channel if we didn't know about it.
    if (not defined $channel) {
        $irc->msg($chan,
            "$who: Sorry, I don't seem to have $chan in my database, which is"
           . " a weird error that needs to be reported!");
       return;
    }

    my $game = $channel->rel_game;

    # How long ago did we last do this?
    my $now = time();

    if (defined $game and defined $self->_last->{$game->id}
            and defined $self->_last->{$game->id}->{status}) {
        my $last_status = $self->_last->{$game->id}->{status};

        if (($now - $last_status) <= 120) {
        # Last time we did status in this channel was 120 seconds ago or less.
            debug("%s tried to display status for %s but it was already done %u"
               . " secs ago; ignoring", $who, $chan, ($now - $last_status));
            $irc->msg($chan,
                sprintf("$who: Sorry, I'm ignoring your status command"
                    . " because I did one just %u secs ago.",
                    ($now - $last_status)));
            return;
        }
    }

    # Record timestamp of when we did this.
    if (defined $game) {
        $self->_last->{$game->id}->{status} = $now;
    }

    if (not defined $game) {
        # There's never been a game in this channel.
        $self->no_such_game($chan, $chan, $who);
    } elsif (2 == $game->status) {
        $self->report_game_status($game, $chan);
    } elsif (1 == $game->status) {
        my $num_players = $game->rel_active_usergames->count;
        my $my_nick     = $irc->nick();

        # Game is still gathering players. Give different response depending on
        # whether they are already in it or not.
        my $ug = $self->db_get_nick_in_game($who, $game);

        if (defined $ug) {
            $irc->msg($chan,
                sprintf("%s: A game exists but we only have %u player%s"
                   . " (%s). Find me %u more and we're on.", $who,
                   $num_players, 1 == $num_players ? '' : 's',
                   1 == $num_players ? 'you' : 'including you',
                   4 - $num_players));
            $irc->msg($chan,
                qq{Any takers? Just type "$my_nick: me" and you're in.});
        } else {
            $irc->msg($chan,
                sprintf("%s: A game exists but we only have %u player%s."
                    . " Find me %u more and we're on.", $who, $num_players,
                    1 == $num_players ? '' : 's', 4 - $num_players));
            $irc->msg($chan,
                qq{$who: How about you? Just type "$my_nick: me" and you're}
                . qq{ in.});
        }
    } elsif (0 == $game->status) {
        $irc->msg($chan,
            "$who: The game is paused but I don't know why! Report this!");
    } else {
        debug("Game for %s has an unexpected status (%u)", $chan,
            $game->status);
        $irc->msg($chan,
            "$who: I'm confused about the state of the game, sorry. Report"
           . " this!");
    }
}

# A user wants to join a (presumably) already-running game. This can happen
# from either of the following scenarios:
#
# <foo> AgainstHumanity: start
# <AgainstHumanity> foo: You're on! We have a game of Perpetually Against
#                   Humanity up in here. 4 players minimum are required. Who
#                   else wants to play?
# <AgainstHumanity> Say "AgainstHumanity: me" if you'd like to!
# <bar> AgainstHumanity: me!
#
# or:
#
# <bar> AgainstHumanity: deal me in.
sub dealin {
    my ($self, $args) = @_;

    my $irc     = $self->_irc;
    my $chan    = $args->{chan};
    my $who     = $args->{nick};
    my $schema  = $self->_schema;
    my $my_nick = $self->_irc->nick();

    my $channel = $self->db_get_channel($chan);

    if (not defined $channel) {
        $self->_irc->msg($chan,
            "$who: I can't seem to find a Channel object for this channel."
           . " That's weird and shouldn't happen. Report this!");
        return;
    }

    my $game = $channel->rel_game;

    debug("%s wants to join game at %s", $who, $chan);

    # Is there a game running already?
    if (not defined $game) {
        # No, there is no game.
        #
        # This raises the question of whether we should treat a user asking to
        # be dealt in to a non-existent game as request to start the game
        # itself.
        #
        # I'm leaning towards "no" because the fact that the channel doesn't
        # already have a game running may hint towards the norms of the channel
        # being that games aren't welcome.
        debug("There's no game at %s for %s to join", $chan, $who);
        $irc->msg($chan,
            "$who: Sorry, there's no game here to deal you in to. Want to start"
           . " one?");
        $irc->msg($chan, qq{$who: If so, type "$my_nick: start"});
        return;
    }

    my $user = $self->db_get_user($who);

    my @active_usergames = $game->rel_active_usergames;

    # Are they already in it?
    if (defined $game->rel_active_usergames
            and grep $_->user == $user->id, @active_usergames) {
        debug("%s is already playing in game at %s", $who, $chan);
        $irc->msg($chan, "$who: Heyyy, you're already playing!");
        return;
    }

    # Is the game's current round complete (waiting on Card Tsar)? If so then no
    # new players can immediately join, because then everyone would know who
    # the extra play was from. They will be dealt in to the next hand.
    if (2 == $game->status and $self->round_is_complete($game)) {
        debug("%s can't immediately join game at %s because the round is"
           . " complete", $who, $chan);

        my $tsar      = $game->rel_tsar_usergame->rel_user;
        my $tsar_nick = do {
            if (defined $tsar->disp_nick) { $tsar->disp_nick }
            else                          { $tsar->nick }
        };

        my $joinq = $self->_joinq;

        $joinq->push(
            {
                user => $user,
                game => $game,
            }
        );

        $irc->msg($chan,
            sprintf("%s: Sorry, this round is complete and we're waiting on"
               . " %s to pick the winner. Stay on IRC and you'll be dealt"
               . " in to the next round automatically.", $who, $tsar_nick));
        return;
    }

    # Maximum 20 players in a game.
    my $num_players = scalar @active_usergames;

    if ($num_players >= 20) {
        debug("%s can't join game at %s because there's already %s players",
            $who, $chan, $num_players);
        $irc->msg($chan,
            "$who: Sorry, there's already $num_players players in this game and"
           . " that's the maximum. Try again once someone has resigned!");
        return;
    }

    # "Let the User see the Game!" Ahem. Add the User to the Game.
    my $usergame = $self->add_user_to_game(
        {
            user => $user,
            game => $game,
        }
    );

    $irc->msg($chan, "$who: Nice! You're in!");

    # Does the game have enough players to start yet?
    $num_players = $game->rel_active_usergames->count;

    if ($num_players >= 4 and 1 == $game->status) {
        debug("Game at %s now has enough players to proceed", $chan);

        $game->status(2);
        $game->activity_time(time());
        $game->update;

        my $prefix;

        if (defined $game->bcardidx) {
            # This game already had some rounds so it has been unpaused.
            $prefix = 'The game is on again!';
        } else {
            # This game has never had any rounds before; this is a new game.
            $prefix = 'The game begins!';
        }

        $irc->msg($chan,
            "$prefix Give me a minute or two to tell everyone their hands"
           . " without flooding myself off, please.");

        # Get a chat window open with all the players.
        $self->brief_players($game);
        # Top everyone's White Card hands up to 10 cards.
        $self->topup_hands($game);
        # And deal out a Black Card to the Tsar, if necessary.
        if (not defined $game->bcardidx) {
            $self->deal_to_tsar($game);
        } else {
            $irc->msg($chan, "Current Black Card:");
            $self->notify_bcard($chan, $game);
        }
    } elsif (1 == $game->status) {
        debug("Game at %s still requires %u more players", $chan,
            4 - $num_players);
        $irc->msg($chan,
            "We've now got $num_players of minimum 4. Anyone else?");
        $irc->msg($chan, qq{Type "$my_nick: me" if you'd like to play too.});
    } elsif (2 == $game->status) {
        # They joined an already-running game, so they need a hand of
        # White Cards.
        $self->topup_hand($usergame);

        # And to know what the Black Card is.
        $irc->msg($who, "Current Black Card:");
        $self->notify_bcard($who, $game);
    }
}

# User wants to start a new game in a channel.
sub start {
    my ($self, $args) = @_;

    my $chan    = $args->{chan};
    my $who     = $args->{nick};
    my $my_nick = $self->_irc->nick();
    my $irc     = $self->_irc;
    my $schema  = $self->_schema;

    # Do we have a channel in the database yet? The only way to create a
    # channel is to be invited there, so there will not be any need to create
    # it here, and it's a weird error to not have it.
    my $channel = $self->db_get_channel($chan);

    if (not defined $channel) {
        $irc->msg($chan,
            "$who: Sorry, I don't seem to have $chan in my database, which is"
           . " a weird error that needs to be reported!");
       return;
    }

    # Is there already a game for this channel?
    my $game = $channel->rel_game;

    if (defined $game) {
        # There's already a Game for this Channel. It could be in one of three
        # possible states:
        #
        # 0: Paused for an unknown reason.
        # 1: Waiting for a sufficient number of players.
        # 2: Running.
        #
        # Whatever the case, this is not the place where it can be started:
        #
        # * Paused games should be started as soon as the bot joins a welcoming
        #   channel.
        #
        # * Games without enough players will start as soon as they get enough
        #   players.
        #
        # * Running games don't need to be started!
        #
        # So apart from explanatory messages this isn't going to do anything.
        my $status = $game->status;

        if (0 == $status) {
            $irc->msg($chan,
                "$who: Sorry, there's already a game for this channel, though"
               . " it seems to be paused when it shouldn't be! Ask around?");
        } elsif (1 == $status) {
            my $count = $game->rel_active_usergames->count;

            $irc->msg($chan,
                "$who: Sorry, there's already a game here but we only have"
               . " $count of minimum 4 players. Does anyone else want to"
               . " play?");
            $irc->msg($chan, qq{Type "$my_nick: me" if you'd like to!});
        } elsif (2 == $status) {
            $irc->msg($chan,
                "$who: Sorry, there's already a game running here!");
        }

        return;
    }

    # Need to create a new Game for this Channel. The User corresponding to the
    # nickname will be its first player. The initial status of the game will be
    # "waiting for players."
    $game = $schema->resultset('Game')->create(
        {
            channel       => $channel->id,
            create_time   => time(),
            activity_time => time(),
            status        => 1,
            packs         => $self->{_config}->{packs},
        }
    );

    # Seems to be necessary in order to get the default DB values back into the
    # object.
    $game->discard_changes;

    # Stuff the cards from memory structure into the database so that this game
    # has its own unique deck to work through, that will persist across process
    # restarts.
    $self->db_populate_cards($game, 'Black');
    $self->db_populate_cards($game, 'White');

    my $user = $self->db_get_user($who);

    # "Let the User see the Game!" Ahem. Add the User to the Game as the Tsar.
    # In the absence of being able to know who pooped last, the starting user
    # will be the first Card Tsar.
    my $usergame = $self->add_user_to_game({
            user => $user,
            game => $game,
            tsar => 1
        }
    );

    # Now tell 'em.
    $irc->msg($chan,
        "$who: You're on! We have a game of Perpetually Against Humanity up in"
       . " here. 4 players minimum are required. Who else wants to play?");
    $irc->msg($chan,
        qq{Say "$my_nick: me" if you'd like to!});
}

# A user wants to resign from the game. If they are the current round's Card
# Tsar then they aren't allowed to resign. Otherwise, their White Cards
# (including any that were already played in this round) are discarded and they
# are removed from the running game.
#
# If this brings the number of players below 4 then the game will be paused.
#
# The player can rejoin the game at a later time.
sub resign {
    my ($self, $args) = @_;

    my $chan    = $args->{chan};
    my $who     = $args->{nick};
    my $schema  = $self->_schema;
    my $my_nick = $self->_irc->nick();
    my $irc     = $self->_irc;

    my $channel = $self->db_get_channel($chan);

    if (not defined $channel) {
        $irc->msg($chan,
            "$who: I can't seem to find a Channel object for this channel."
           . " That's weird and shouldn't happen. Report this!");
        return;
    }

    my $game = $channel->rel_game;

    debug("%s attempts to resign from game in %s", $who, $chan);

    # Is there a game actually running?
    if (not defined $game) {
        debug("%s can't resign from non-existent game in %s", $who, $chan);
        $irc->msg($chan, "$who: There isn't a game running at the moment.");
        return;
    }

    my $user = $self->db_get_user($who);

    my $usergame = $schema->resultset('UserGame')->find(
        {
            'user' => $user->id,
            'game' => $game->id,
        },
    );

    # Is the user active in the game?
    if (not defined $usergame or 0 == $usergame->active) {
        # No.
        debug("%s tried to resign from game in %s but they weren't active",
            $who, $chan);
        $irc->msg($chan, "$who: You're not playing!");
        return;
    }

    $irc->msg($chan, "$who: Okay, you've been dealt out of the game.");
    $irc->msg($chan,
        qq{$who: If you want to join in again later then type}
        . qq{ "$my_nick: deal me in"});

    $self->resign($usergame);
}

# User wants to pick a winning play.
sub winner {
    my ($self, $args) = @_;

    my $irc     = $self->_irc;
    my $chan    = $args->{chan};
    my $who     = $args->{nick};
    my $winner  = $args->{params};
    my $schema  = $self->_schema;
    my $my_nick = $self->_irc->nick();

    # winner needs to exist and be a single positive integer.
    if (not defined $winner or $winner !~ /^\d+$/ or $winner < 1) {
        $irc->msg($chan,
            qq{$who: What? You need to give me a single number, e.g.}
           . qq{ "$my_nick: winner 1"});
       return;
   }

    my $channel = $self->db_get_channel($chan);

    if (not defined $channel) {
        $irc->msg($chan,
            "$who: I can't seem to find a Channel object for this channel."
            . " That's weird and shouldn't happen. Report this!");
        return;
    }

    my $game = $channel->rel_game;

    # Is there even a game?
    if (not defined $game) {
        $irc->msg($chan, "$who: What? There's no game running here yet.");
        return;
    }

    # Is the game active?
    if ($game->status != 2) {
        $irc->msg($chan, "$who: Sorry, the game isn't active right now.");

        # Maybe it needs more players.
        if (1 == $game->status) {
            my $num_players = $game->rel_active_usergames->count;

            $irc->msg($chan,
                sprintf("We need %u more player%s before we can start"
                   . " playing.", 4 - $num_players,
                   (4 - $num_players) == 1 ? '' : 's'));
        }

        return;
    }

    # Are they the Card Tsar?
    my $user = $self->db_get_user($who);

    if (not $self->user_is_tsar($user, $game)) {
        my $tsar      = $game->rel_tsar_usergame->rel_user;
        my $tsar_nick = $tsar->disp_nick;

        $tsar_nick = $tsar->nick if (not defined $tsar_nick);

        $irc->msg($chan,
            sprintf("%s: Sorry, you're not the Card Tsar – that's %s.",
                $who, $tsar_nick));
        return;
    }

    # Is the round actually complete?
    if (not $self->round_is_complete($game)) {
        $irc->msg($chan,
            "$who: Sorry, not everyone has played their hand yet!");
        $irc->msg($chan, $self->build_waitstring($game));
        return;
    }

    my $tally = $self->_plays->{$game->id};

    foreach my $uid (keys %{ $tally }) {
        if ($tally->{$uid}->{seq} == $winner) {
            # Found the winning play.
            my $winuser = $schema->resultset('User')->find(
                {
                    id => $uid,
                }
            );

            $game->rel_tsar_usergame->activity_time(time());
            $game->rel_tsar_usergame->update;

            my $win_ug = $self->end_round($winuser, $game);
            $self->cleanup_plays($game);
            $self->pick_new_tsar($win_ug, $tally->{$uid}->{play}, $game);
            $self->topup_hands($game);
            $self->clear_pokes($game);
            return;
        }
    }

    # If we got here then they gave us a winner number that doesn't exist in
    # the plays.
    $irc->msg($chan,
        "$who: Sorry, I don't seem to have a record of a play with that"
       . " number.");
}

# User wants to call up a list of plays for the completed hand.
sub plays {
    my ($self, $args) = @_;

    my $irc     = $self->_irc;
    my $chan    = $args->{chan};
    my $who     = $args->{nick};
    my $my_nick = $irc->nick();

    my $channel = $self->db_get_channel($chan);

    # It shouldn't be possible to not have a Channel row, because we wouldn't
    # be inside the channel if we didn't know about it.
    if (not defined $channel) {
        $irc->msg($chan,
            "$who: Sorry, I don't seem to have $chan in my database, which is"
            . " a weird error that needs to be reported!");
        return;
    }

    my $game = $channel->rel_game;

    # How long ago did we last do this?
    my $now = time();

    if (defined $game and defined $self->_last->{$game->id}
            and defined $self->_last->{$game->id}->{plays}) {
        my $last_plays = $self->_last->{$game->id}->{plays};

        if (($now - $last_plays) <= 120) {
            # Last time we did plays in this channel was 120 seconds ago or
            # less.
            debug("%s tried to display plays for %s but it was already done %u"
                . " secs ago; ignoring", $who, $chan, ($now - $last_plays));
            $irc->msg($chan,
                sprintf("$who: Sorry, I'm ignoring your plays command"
                    . " because I did one just %u secs ago.",
                    ($now - $last_plays)));
            return;
        }
    }

    # Record timestamp of when we did this.
    if (defined $game) {
        $self->_last->{$game->id}->{plays} = $now;
    }

    if (not defined $game) {
        # There's never been a game in this channel.
        $self->no_such_game($chan, $chan, $who);
    } else {
        $self->report_plays($game, $chan);
    }
}

1;
